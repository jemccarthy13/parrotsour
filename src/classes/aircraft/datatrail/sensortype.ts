export enum SensorType {
  RAW,
  ARROW,
}
