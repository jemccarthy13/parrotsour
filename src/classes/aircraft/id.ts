export enum IDMatrix {
  HOSTILE = "red",
  SUSPECT = "orange",
  ASSUME_FRIEND = "green",
  NEUTRAL = "purple",
  FRIEND = "blue",
}

export function getMostRestrictiveID(ids: IDMatrix[]): IDMatrix {
  function idToNum(id: IDMatrix) {
    const num = 4
    switch (id) {
      case IDMatrix.SUSPECT:
        return 3
      case IDMatrix.NEUTRAL:
        return 2
      case IDMatrix.ASSUME_FRIEND:
        return 1
      case IDMatrix.FRIEND:
        return 1
    }
    return num
  }
  let lowestID = IDMatrix.HOSTILE
  let lowestIDNum = 4
  for (let i = 0; i < ids.length; i++) {
    const idNum = idToNum(ids[i])
    if (idNum < lowestIDNum) {
      lowestIDNum = idNum
      lowestID = ids[i]
    }
  }
  return lowestID
}
