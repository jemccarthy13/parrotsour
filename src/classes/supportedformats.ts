/**
 * This file contiains a list of format options.
 * CAO 2021 - it's "ALSA" formatting, or "IPE" (3-3 IPE)
 */
/**
 * List of available formats.
 */
export enum FORMAT {
  ALSA,
  IPE,
  CLOSE,
}
