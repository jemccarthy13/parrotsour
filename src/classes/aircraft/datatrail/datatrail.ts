import { Point } from "../../point"
import { IDMatrix } from "../id"

// TODO -- DATATRAIL -- register? so that Aircraft can generate one of each type of datatrail?
// TODO -- DATATRAIL -- Support a toggle for the different Sensor Types.....
export abstract class DataTrail {
  private startPos: Point

  constructor(startPos: Point) {
    this.startPos = startPos
  }

  public getStartPos(): Point {
    return this.startPos
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public move(_heading: number): void {
    //console.warn("Moving DataTrail in base class. Will have no effect.")
  }

  abstract getCenterOfMass(heading: number): Point

  abstract draw(heading: number, id: IDMatrix): void
}
