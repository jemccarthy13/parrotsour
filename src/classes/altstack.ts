import { FORMAT } from "./supportedformats"

export interface AltStack {
  stack: string
  fillIns: string
}

/**
 * Non-exported function to format alt stacks and fill-ins for group altitudes
 *
 * A single group will be no-op (return hard alt + fillin if HIGH)
 * A group without stacks will return single alt + fillin if HIGH)
 * Otherwise, will return highest altitude for each "bucket" and # contacts hi/med/low
 *
 * @param altitudes - group's altitudes for each contact
 * @param format - comm format
 */
export function getAltStack(altitudes: number[], format: FORMAT): AltStack {
  // convert altitudes to 3-digit flight level and sort low->high
  const formattedAlts: string[] = altitudes
    .map((a: number) => ("0" + a).slice(-2) + "0")
    .sort()
    .reverse()

  const stackHeights: string[] = []
  let stackIndexes: number[] = []

  // break out into bins of 10k foot separation between contacts
  for (let x = formattedAlts.length; x >= 0; x--) {
    const diff: number =
      parseInt(formattedAlts[x - 1]) - parseInt(formattedAlts[x])
    if (diff >= 100) {
      stackHeights.push(formattedAlts[x])
      stackIndexes.push(x)
    }
  }

  stackIndexes = stackIndexes.reverse()

  // get the highest altitude within each bucket for formatting
  const stacks: string[][] = []
  let lastZ = 0
  for (let z = 0; z < stackIndexes.length; z++) {
    stacks.push(formattedAlts.slice(lastZ, stackIndexes[z]))
    lastZ = stackIndexes[z]
  }
  stacks.push(formattedAlts.slice(lastZ))

  // format to "##k"
  let answer = formattedAlts[0].replace(/0$/, "k") + " "
  let answer2 = ""

  // do formatting

  // if no stack, look for >40k for "HIGH"
  if (stacks.length <= 1) {
    altitudes.sort()
    if (altitudes[altitudes.length - 1] >= 40) {
      answer2 += " HIGH "
    }
    // otherwise, print stacks
  } else {
    answer = "STACK "
    for (let y = 0; y < stacks.length; y++) {
      // check to add "AND" for alsa, when on last stack alt
      const AND = y === stacks.length - 1 && format !== FORMAT.IPE ? "AND " : ""
      answer += AND + stacks[y][0].replace(/0$/, "k") + " "
    }

    // format # hi/med/low when there are at least 3 contacts
    // if there are 3 contacts and 3 altitudes, 1 hi / 1 med / 1 low is not required
    // (so skip this)
    if (
      altitudes.length > 2 &&
      !(altitudes.length === stacks.length && stacks.length === 3)
    ) {
      switch (stacks.length) {
        case 2:
          answer2 += stacks[0].length + " HIGH "
          answer2 += stacks[1].length + " LOW "
          break
        case 3:
          answer2 += stacks[0].length + " HIGH "
          answer2 += stacks[1].length + " MEDIUM "
          answer2 += stacks[2].length + " LOW "
          break
      }
    }
  }

  return { stack: answer.trim(), fillIns: answer2.trim() }
}
