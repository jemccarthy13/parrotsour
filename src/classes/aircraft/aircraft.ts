// Classes and interfaces
import { DataTrail } from "./datatrail/datatrail"
import { SensorType } from "./datatrail/sensortype"
import { GroupParams } from "../groups/group"
import { IDMatrix } from "./id"
import { AircraftIntent, IntentParams } from "./intent"
import { Point } from "../point"
import Tasking from "../taskings/tasking"

// Functions
import {
  getDegDeltaBetween,
  headingToRadians,
  randomNumber,
} from "../../utils/psmath"
import { DataTrailFactory } from "./datatrail/datatrailfactory"

/*
 * TODO -- EWI -- for fill-ins, implement more AC types?
 */
export enum ACType {
  FTR,
  RPA,
  TANK,
  AEW,
}

/**
 * Aircraft have all the parameters of Groups, plus a single
 * altitude and configurable type.
 */
interface AircraftParams extends GroupParams {
  alt: number
  type: ACType
}

/**
 * TODO -- AIRCRAFT -- consider moving to a structure like
 * https://khalilstemmler.com/blogs/typescript/getters-and-setters/
 * To reduce the number of functions provided in the aircraft "api"
 */
export class Aircraft {
  // AASI (alt, aspect, speed, ID)
  private altitude: number
  private heading: number
  // TODO -- SPEED -- implement aircraft speed logic
  // private speed: number
  private id: IDMatrix

  private dataTrail: Map<number, DataTrail>

  private type: ACType

  private startPos = new Point(0, 0)

  private intent = new AircraftIntent()
  private tasking: Tasking | undefined
  private capping = false

  private ctx: CanvasRenderingContext2D | undefined

  constructor(p?: Partial<AircraftParams>) {
    if (!p) p = {}

    // Take given heading/ID/EWI or use default
    this.heading = p.hdg || 90
    this.id = p.id || IDMatrix.HOSTILE
    this.type = p.type || ACType.FTR

    // Set current position
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // TODO -- DATATRAIL -- store a map of SensorType -> DataTrail
    // here, so it has both kinds of data trails available.
    // Then, in the PSControls, toggling data trail types only
    // triggers a redraw of existing groups' data trails (via .draw())
    this.dataTrail = new Map<number, DataTrail>()

    for (const type in SensorType) {
      if (!Number.isNaN(Number(type)))
        this.dataTrail.set(
          Number(type),
          DataTrailFactory.create(
            Number(type),
            this.startPos,
            this.getHeading()
          )
        )
    }

    // Set current altitude
    let low = 15
    let hi = 45
    if (this.type === ACType.RPA) {
      low = 0o5
      hi = 18
    }
    this.altitude = p.alt || randomNumber(low, hi)

    // Current matches desired initially
    this.intent.setDesiredHeading(this.heading)
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this Aircraft's draw will start; will be approximately
   * the 'rear'' end of the data trail.
   */
  public getStartPos(): Point {
    return this.startPos
  }

  /**
   * Get the center of mass position for this group.
   *
   * The type of data trail dictates where the center of mass is.
   * (i.e. radar - last sensed radar vs. arrow - loc of arrow head)
   *
   * @param dataStyle (Radar & IFF) | Arrows - must explicitly be set if drawing
   *                  anything other than arrows
   * @param ctx Current drawing context
   */
  getCenterOfMass(dataStyle?: SensorType): Point {
    if (this.isCapping()) {
      return this.getStartPos()
    }
    dataStyle = dataStyle === undefined ? SensorType.ARROW : dataStyle
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      return dataTrail.getCenterOfMass(this.getHeading())
    } else {
      return new Point(-1, -1)
    }
  }

  /**
   * @returns true iff Aircraft is capping (at desired final Route location)
   */
  isCapping(): boolean {
    return this.capping
  }

  /**
   * @param newVal New value for isCapping
   */
  setCapping(newVal: boolean): void {
    this.capping = newVal
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/
  /**
   * @returns Current heading of this Aircraft
   */
  public getHeading(): number {
    return this.heading
  }

  /**
   * @param hdg Set new heading of this Aircraft
   */
  public setHeading(hdg: number): void {
    this.heading = hdg
  }

  /*************************************************************************
   * Altitude
   *************************************************************************/
  /**
   * @returns Current altitude of this Aircraft
   */
  public getAltitude(): number {
    return this.altitude
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current IDMatrix (HOS/NEU/etc) of this Aircraft.
   */
  getIDMatrix(): IDMatrix {
    return this.id
  }
  /**
   * @param newID Set new IDMatrix ID for this Aircraft
   */
  setIDMatrix(newID: IDMatrix): void {
    this.id = newID
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Current Aircraft type
   */
  getType(): ACType {
    return this.type
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw the Aircraft, with it's associated data trail and symbology
   * @param context  Current drawing style
   * @param dataStyle The SensorType of the DataTrail
   */
  draw(dataStyle: SensorType): void {
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      dataTrail.draw(this.getHeading(), this.getIDMatrix())
    }
  }

  /**
   * Move the arrow once based on the current heading / vector.
   */
  move(): void {
    if (!this.capping) {
      // don't move if capping at final destination
      // calculate & apply offsets based on start/end position
      const rads: number = headingToRadians(this.getHeading()).radians
      const offsetX: number = 7 * Math.cos(rads)
      const offsetY: number = -7 * Math.sin(rads)
      this.startPos.x += offsetX
      this.startPos.y += offsetY

      this.dataTrail.forEach((value: DataTrail) => {
        value.move(this.getHeading())
      })

      // apply a turn as required
      this.turnToTarget()
    }
  }

  /**
   * Make a logical turn towards the Aircraft's target by applying an offset
   * to the current heading.
   */
  turnToTarget(): void {
    // if Aircraft wants to go somewhere
    // determine the bearing to the next route point
    const tgtPos = this.intent.getNextRoutingPoint()
    if (tgtPos) {
      this.intent.setDesiredHeading(
        this.getCenterOfMass().getBR(tgtPos).bearingNum
      )
    }

    // how "far away" from current heading is the desired?
    const turnDegrees = getDegDeltaBetween(
      this.getHeading(),
      this.intent.getDesiredHeading(),
      this.intent.getForcedTurn()
    )

    // apply 'logical' turn by only turning so far towards the target
    // each frame
    let divisor = 7
    const absDelt = Math.abs(turnDegrees)
    if (absDelt >= 90) {
      divisor = 15
    } else if (absDelt < 7) {
      divisor = 1
    }

    this.setHeading(Math.floor(this.getHeading() + turnDegrees / divisor))
  }

  /**
   * If not at desired altitude, increment to climb/descend towards intended altitude.
   */
  doNextAltChange(): void {
    const atDesiredAlt = this.intent.getDesiredAltitude() === this.getAltitude()
    const CHANGE_PER_FRAME = 0.5
    if (!atDesiredAlt) {
      if (this.intent.getDesiredAltitude() > this.getAltitude()) {
        this.altitude += CHANGE_PER_FRAME // climb
      } else {
        this.altitude -= CHANGE_PER_FRAME // descend
      }
    }
  }

  /*************************************************************************
   * Routing & Intent
   *************************************************************************/
  /**
   *
   * @param newIntent Update this aircraft's intent based on Params
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.intent.updateIntent(newIntent)
  }

  /**
   * @returns true iff there are no desired locations left in the intended route
   */
  atFinalDestination(): boolean {
    return this.intent.atFinalDestination()
  }

  /**
   * @returns Point, if there's a next routing point, or
   * undefined if no intended destination set
   */
  getNextRoutingPoint(): Point | undefined {
    return this.intent.getNextRoutingPoint()
  }

  /**
   * @param pt A new point to be added to the Aircraft's route. It is added to the end
   * of the list.
   */
  addRoutingPoint(pt: Point): void {
    this.intent.addRoutingPoint(pt)
  }

  /**
   * Check current against intended, modify heading and intent as necessary.
   */
  doNextRouting(): void {
    if (!this.intent.atFinalDestination()) {
      // Aircraft still has a route but we've reached our next destination
      // so pop the first point out of the intended route
      if (this.intent.atNextRoutingPoint(this.getStartPos())) {
        this.intent.removeRoutingPoint()
      }
      // If Aircraft still has a point left in routing, update desired heading
      // and turn towards it
      const nextPt = this.intent.getNextRoutingPoint()
      if (nextPt)
        this.updateIntent({
          desiredHeading: this.getCenterOfMass().getBR(nextPt).bearingNum,
        })
      this.turnToTarget()
    } else {
      // Aircraft has reached final destination, so if intent matches it is capping
      this.capping = this.intent.atFinalDestination()
    }
  }

  clearRouting(): void {
    this.intent.clearRouting()
  }

  /*************************************************************************
   * Taskings
   *************************************************************************/
  /**
   * @returns true iff Aircraft has a tasking assigned.
   */
  isTasked(): boolean {
    return this.tasking !== undefined
  }

  /**
   * Clears the current tasking.
   */
  clearTasking(): void {
    this.tasking = undefined
  }

  /**
   * @param task Set a new Tasking for this Aircraft.
   */
  setTasking(task: Tasking): void {
    this.tasking = task
  }

  /**
   * @returns The current Tasking, or undefined if Aircraft has none.
   */
  getTasking(): Tasking | undefined {
    return this.tasking
  }
}
