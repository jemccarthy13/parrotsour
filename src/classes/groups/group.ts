// Classes & Interfaces
import { SensorType } from "../aircraft/datatrail/sensortype"
import { Point } from "../point"
import { ACType, Aircraft } from "../aircraft/aircraft"
import { getMostRestrictiveID, IDMatrix } from "../aircraft/id"
import { IntentParams } from "../aircraft/intent"
import Tasking from "../taskings/tasking"
import { FORMAT } from "../supportedformats"

import { AltStack, getAltStack } from "../altstack"
// Functions
import {
  headingToRadians,
  PIXELS_TO_NM,
  randomNumber,
} from "../../utils/psmath"
import { drawGroupCap } from "./groupcap"
import { Braaseye } from "../braaseye"
import { Aspect, aspectFromCATA, toCardinal } from "../../utils/aspect"
import { BRAA } from "../braa"
import RangeBack from "./rangeback"

/**
 * The types of data that can be used to seed a group.
 *
 * Everything not provided is randomized or defaulted.
 */
export interface GroupParams {
  dataTrailType: SensorType
  sx?: number
  sy?: number
  nContacts?: number
  hdg?: number
  alts?: number[]
  desiredHdg?: number
  id?: IDMatrix
  type?: ACType
}

/**
 * An AircraftGroup is an array of 1+ Aircraft.
 *
 * It contains logic to move, give intent to, and draw Aircraft.
 *
 * TODO -- GROUP SPLIT/MERGE -- Currently maneuvers are handled as an all
 * or nothing. In the animator, consider additional logic for (man > 0), to update
 * "groups" var to split groups. Should be 'easy'
 */
export class AircraftGroup extends Array<Aircraft> {
  private braaseye!: Braaseye

  private startPos: Point = Point.DEFAULT
  private label = "GROUP"
  private useTrackDir = true
  private maneuvers = 0
  private anchor = false
  private useBull = false

  /**
   * Construct an AircraftGroup from the given parameters.Group will initialize
   * what it can, randomize what it can, and default the rest.
   *
   * @param p A Partial object of GroupParams.
   */
  constructor(p?: Partial<GroupParams>) {
    super()
    if (!p) p = {}
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // Seed params to pass into Aircraft
    p.sx = this.startPos.x
    p.sy = this.startPos.y
    p.hdg = p.hdg || 90

    this.startPos = new Point(p.sx, p.sy)

    // Create nContacts number of Aircraft and add to this collection
    const nContacts = p.nContacts || randomNumber(1, 4)
    for (let contact = 0; contact < nContacts; contact++) {
      if (p.alts && p.alts[contact])
        this.push(new Aircraft({ ...p, alt: p.alts[contact] }))
      else {
        this.push(new Aircraft(p))
      }

      // Compute start position offset for follow-on groups (calculated from
      // 90 deg perpendicular)
      const vectors = headingToRadians(p.hdg)
      p.sx += 2 * PIXELS_TO_NM * Math.cos(vectors.offset)
      p.sy += 2 * PIXELS_TO_NM * -Math.sin(vectors.offset)
    }

    // TODO -- MANEUVER -- % chance of 2 maneuvers (i.e. flank turn back hot)
    if (randomNumber(0, 100) < 20) {
      this.maneuvers = 1
    }
  }

  /*************************************************************************
   * Formatting
   *************************************************************************/
  setAnchor(b: boolean): void {
    this.anchor = b
    this.useBull = b
  }

  isAnchor(): boolean {
    return this.anchor
  }

  setUseBull(b: boolean): void {
    this.useBull = b
  }

  getUseBull(): boolean {
    return this.useBull
  }

  formatNumContacts(): string {
    let answer = ""
    if (this.getStrength() > 1) {
      if (this.getStrength() >= 3) {
        answer += "HEAVY "
      }
      answer += this.getStrength() + " CONTACTS"
    }
    return answer
  }

  /**
   * Return the string formatted answer for this group based on properties of the group
   */
  format(format: FORMAT, rangeBack?: RangeBack): string {
    // format label
    let answer = this.getLabel() + " "

    // format separation
    if (rangeBack) {
      answer += rangeBack.toString() + " "
    }

    // format bullseye if anchor priority
    const braaseye = this.getBraaseye()
    if (this.useBull || false) {
      answer += "BULLSEYE " + braaseye.bull.toString() + ", "
    }

    // format altitude stack
    const altStack = this.getAltStack(format)
    answer += altStack.stack + " "

    // format track direction
    const trackDir = this.getTrackDir()
    answer += " "
    answer += trackDir !== undefined ? trackDir : ""

    // apply ID
    answer += " HOSTILE "

    // apply fill-in for # contacts
    answer += this.formatNumContacts() + " "

    // apply fill-ins (HI/FAST/etc)
    answer += " " + altStack.fillIns
    return answer.replace(/ {2}/g, " ")
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this group's draw will start; will be approximately
   * the 'rear' end of the data trail.
   */
  getStartPos(): Point {
    return this[0].getStartPos()
  }

  /**
   * "Center of Mass" for an aircraft/group is one projected data plot ahead of the
   * data trail, averaged. For Aircraft, this is equiv. to one projected data plot ahead
   * of the Aircraft, on it's current heading.
   *
   * @returns The "center of mass" for this group, which is the average of each
   * Aircraft's center of mass.
   */
  getCenterOfMass(dataStyle: SensorType): Point {
    if (this.isCapping()) {
      return this.startPos
    }
    let x = 0
    let y = 0
    for (let idx = 0; idx < this.getStrength(); idx++) {
      const acPos = this[idx].getCenterOfMass(dataStyle)
      x += acPos.x
      y += acPos.y
    }
    return new Point(x / this.getStrength(), y / this.getStrength())
  }

  /**
   * @returns true iff one Aircraft in this group is capping
   */
  isCapping(): boolean {
    return (
      this.find((ac) => {
        return !ac.isCapping()
      }) === undefined
    )
  }

  /**
   * Set each aircraft in this group to be capping
   */
  setCapping(newVal: boolean): void {
    this.forEach((ac) => ac.setCapping(newVal))
  }

  setBraaseye(braaseye: Braaseye): void {
    this.braaseye = braaseye
  }

  getBraaseye(): Braaseye {
    return this.braaseye
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/

  /**
   * Get 'aspect' (HOT/FLANK/BEAM, etc) to another group
   *
   * Aspect is calculated by taking the angle difference between
   * other a/c heading, and the reciprocal bearing between this group
   * and other a/c.
   *
   * @param otherGrp other aircraft
   */
  getAspect(otherGrp: AircraftGroup, dataStyle: SensorType): Aspect {
    const recipBrg: BRAA = otherGrp
      .getCenterOfMass(dataStyle)
      .getBR(this.getCenterOfMass(dataStyle))

    let dist = (otherGrp.getHeading() - parseInt(recipBrg.bearing) + 360) % 360
    if (dist > 180) dist = 360 - dist
    const cata = dist

    return aspectFromCATA(cata)
  }

  /**
   * @returns Current heading of the Group. Currently assumes all Aicraft are
   * following the same heading, therefore only returns first Aircraft's heading.
   */
  getHeading(): number {
    return this[0].getHeading()
  }

  /**
   * @param newHdg New heading to set for all aircraft in this group
   */
  setHeading(newHdg: number): void {
    this.forEach((ac) => ac.setHeading(newHdg))
  }

  /**
   * @returns Cardinal track direction or undefined. If there is a picture
   * track direction specified (i.e. all groups track West) this function
   * returns undefined. Otherwise, cardinal direction from heading.
   */
  getTrackDir(): string | undefined {
    let trackDir = undefined
    if (this.isCapping()) {
      trackDir = "CAP"
    } else if (this.useTrackDir) {
      trackDir = "TRACK " + toCardinal(this.getHeading())
    }
    return trackDir
  }

  /**
   * Set the useTrackDir flag, whether or not to include track direction
   * in group formatting
   */
  setUseTrackDir(newVal: boolean): void {
    this.useTrackDir = newVal
  }

  /**
   * Update maneuver count
   * @param newNumManeuvers New number of maneuvers remaining
   */
  setManeuvers(newNumManeuvers: number): void {
    this.maneuvers = newNumManeuvers
  }

  /**
   * @returns true iff group has at least one maneuver left.
   */
  doesManeuvers(): boolean {
    return this.maneuvers > 0
  }

  /*************************************************************************
   * Altitude(s)
   *************************************************************************/
  /**
   * @returns Altitude of the highest Aircraft in this group
   */
  getAltitude(): number {
    return Math.max(...this.getAltitudes())
  }

  /**
   * @param format The desired formatting
   * @returns AltStack containing formatted STACKS/Fillins
   */
  getAltStack(format: FORMAT): AltStack {
    return getAltStack(this.getAltitudes(), format)
  }

  /**
   * @returns number[] containing the altitude of every Aircraft in this Group
   */
  getAltitudes(): number[] {
    const alts = []
    for (let idx = 0; idx < this.length; idx++) {
      alts.push(this[idx].getAltitude())
    }
    return alts
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current group label
   */
  getLabel(): string {
    return this.label
  }

  /**
   * @param newLbl New group label
   */
  setLabel(newLbl: string): void {
    this.label = newLbl
  }

  /**
   * Update group ID
   * @param newID
   */
  setIDMatrix(newID: IDMatrix): void {
    this.forEach((ac) => {
      ac.setIDMatrix(newID)
    })
  }

  /**
   * @returns The most restrictive ID of child Aircraft
   */
  getIDMatrix(): IDMatrix {
    const ids: IDMatrix[] = []
    this.forEach((ac) => {
      ids.push(ac.getIDMatrix())
    })
    return getMostRestrictiveID(ids)
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Number of contacts in this group
   */
  getStrength(): number {
    return this.length
  }

  /**
   * @returns The Type of Aircraft in this group. Currently only supports one type
   * for the entire group (first aircraft)
   */
  getType(): ACType {
    return this[0].getType()
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw each aircraft. See Aircraft.draw(...)
   * @param context Current drawing context
   * @param dataStyle The type of DataTrail to use
   */
  draw(dataStyle: SensorType): void {
    if (this.isCapping()) {
      drawGroupCap(this)
    } else {
      this.forEach((ac) => ac.draw(dataStyle))
    }
  }

  /**
   * Move each Aircraft. See Aircraft.move()
   */
  move(): void {
    this.forEach((ac) => ac.move())
  }

  /**
   * Update each Aircraft's altitude based on intent. See Aircraft.doNextAltChange()
   */
  updateAltitude(): void {
    this.forEach((ac) => ac.doNextAltChange())
  }

  /*************************************************************************
   * Routing
   *************************************************************************/
  /**
   * Update intent for each Aircraft in the group. See Aircraft.updateIntent()
   * @param newIntent Object representing new intent for each Aircraft.
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.forEach((ac) => ac.updateIntent(newIntent))
  }

  /**
   * @returns true iff there is an Aircraft with intended destination
   */
  hasRouting(): boolean {
    return this.find((ac) => !ac.atFinalDestination()) !== undefined
  }

  /**
   * @returns The next Point on this group's intended route. Assumes
   * all Aircraft in the Group are following the same route, therefore returns
   * only the first Aircraft's next routing point.
   */
  getNextRoutingPoint(): Point | undefined {
    return this[0].getNextRoutingPoint()
  }

  /**
   * @param pt Point to add to all group Aircraft(s) routes
   */
  addRoutingPoint(pt: Point): void {
    this.forEach((ac) => ac.addRoutingPoint(pt))
  }

  /**
   * Perform the next routing action for each aircraft. See Aircraft.doNextRouting()
   */
  doNextRouting(): void {
    this.forEach((ac) => ac.doNextRouting())
  }

  /**
   * Clears all currently assigned transit locations
   */
  clearRouting(): void {
    this.forEach((ac) => ac.clearRouting())
  }

  /*************************************************************************
   * Tasking
   *************************************************************************/
  /**
   * @param task Aircraft's new Tasking.
   */
  setTasking(task: Tasking): void {
    this.forEach((ac: Aircraft) => {
      ac.setTasking(task)
    })
  }

  /**
   * @returns true if there is at least one Aircraft in this group that has a Tasking.
   */
  isOnTask(): boolean {
    return this.find((ac) => ac.isTasked()) !== undefined
  }
}
