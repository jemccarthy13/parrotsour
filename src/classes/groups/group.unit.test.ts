import { Point } from "../point"
import * as AltStackHelper from "../altstack"
import { AircraftGroup } from "./group"
import { ACType, Aircraft } from "../aircraft/aircraft"
import { IDMatrix } from "../aircraft/id"
import { SensorType } from "../aircraft/datatrail/sensortype"
import Tasking from "../taskings/tasking"
import { FORMAT } from "../supportedformats"
import { PaintBrush } from "../../canvas/draw/paintbrush"

/**
 * TODO -- TESTING -- Underlying random utility
 * https://luetkemj.github.io/170421/mocking-modules-in-jest
 */
describe("AircraftGroup", () => {
  let canvas: HTMLCanvasElement
  beforeAll(() => {
    canvas = document.createElement("canvas")
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const ctx = canvas.getContext("2d")!
    PaintBrush.use(ctx)
  })

  describe("accessor_functions", () => {
    it("constructor_and_aircraft_functions", () => {
      const startPt = new Point(20, 20)
      const grp = new AircraftGroup({
        sx: startPt.x,
        sy: startPt.y,
        hdg: 90,
        nContacts: 2,
        alts: [100, 200],
      })
      grp.updateIntent({ desiredHeading: 90 })
      expect(grp.getStartPos()).toEqual(startPt)
      expect(grp.isCapping()).toEqual(false)
      grp.doNextRouting()
      expect(grp.isCapping()).toEqual(true)

      grp.addRoutingPoint(new Point(50, 50))
      expect(grp.hasRouting()).toEqual(true)
    })

    describe("location", () => {
      let counter = 0
      jest
        .spyOn(Aircraft.prototype, "getCenterOfMass")
        .mockImplementation(() => {
          counter += 10
          return new Point(50, 50 + counter)
        })
      const grp = new AircraftGroup({ nContacts: 2 })
      expect(grp.getCenterOfMass(SensorType.ARROW)).toEqual(new Point(50, 65))
    })

    describe("altitude", () => {
      it("group_altitudes", () => {
        const grp = new AircraftGroup({
          nContacts: 2,
          alts: [100, 200],
        })
        expect(grp.getAltitude()).toEqual(200)
        expect(grp.getAltitudes()).toEqual([100, 200])
      })

      it("alt_stack", () => {
        const grp = new AircraftGroup({
          nContacts: 2,
          alts: [100, 200],
        })

        expect(grp.getAltitude()).toEqual(200)
        expect(grp.getAltitudes()).toEqual([100, 200])

        const myMock = jest
          .spyOn(AltStackHelper, "getAltStack")
          .mockImplementationOnce(jest.fn())
        grp.getAltStack(FORMAT.ALSA)
        expect(myMock).toHaveBeenCalled()
      })
    })

    describe("id_and_labeling", () => {
      it("allows_labeling", () => {
        const grp = new AircraftGroup()
        expect(grp.getLabel()).toEqual("GROUP")
        grp.setLabel("WEST")
        expect(grp.getLabel()).toEqual("WEST")
      })

      it("tracks_id_matrix", () => {
        const grp = new AircraftGroup({
          nContacts: 4,
        })
        grp.setIDMatrix(IDMatrix.HOSTILE)
        expect(grp.getIDMatrix()).toEqual(IDMatrix.HOSTILE)
        grp[1].setIDMatrix(IDMatrix.SUSPECT)
        expect(grp.getIDMatrix()).toEqual(IDMatrix.SUSPECT)
        grp[2].setIDMatrix(IDMatrix.ASSUME_FRIEND)
        expect(grp.getIDMatrix()).toEqual(IDMatrix.ASSUME_FRIEND)
        grp[2].setIDMatrix(IDMatrix.NEUTRAL)
        expect(grp.getIDMatrix()).toEqual(IDMatrix.NEUTRAL)
      })
    })

    it("track_dir_functions", () => {
      const grp = new AircraftGroup({
        hdg: 90,
      })
      expect(grp.getStrength()).toBeGreaterThanOrEqual(1)
      grp.updateIntent({ desiredHeading: 90 })
      expect(grp.getTrackDir()).toEqual("TRACK EAST")

      grp.setUseTrackDir(false)
      expect(grp.getTrackDir()).toEqual(undefined)

      grp.doNextRouting()
      expect(grp.isCapping()).toEqual(true)
      expect(grp.getTrackDir()).toEqual("CAP")
    })
  })

  describe("movement_functions", () => {
    it("access_maneuvers", () => {
      const grp = new AircraftGroup({ nContacts: 4 })
      expect(grp.setManeuvers(0))
      expect(grp.doesManeuvers()).toEqual(false)
    })

    it("moves_all_aircraft", () => {
      const grp = new AircraftGroup({ nContacts: 4 })
      const myMock = jest
        .spyOn(Aircraft.prototype, "move")
        .mockImplementationOnce(jest.fn())
      grp.move()
      expect(myMock).toHaveBeenCalledTimes(4)
    })
  })

  describe("draw_group", () => {
    it("draws_all_groups", () => {
      const grp = new AircraftGroup({
        nContacts: 4,
        hdg: 135,
        id: IDMatrix.FRIEND,
      })
      const myMock = jest.spyOn(Aircraft.prototype, "draw")
      grp.draw(SensorType.ARROW)
      expect(myMock).toHaveBeenCalledTimes(4)
      expect(canvas).toMatchSnapshot()
    })
  })

  describe("passthrough_functions", () => {
    it("update_alt_each_aircraft", () => {
      const grp = new AircraftGroup({
        nContacts: 2,
        hdg: 135,
        id: IDMatrix.FRIEND,
        alts: [20, 20],
      })
      grp.updateIntent({ desiredAlt: 40 })
      grp.updateAltitude()
      expect(grp[0].getAltitude()).toEqual(20.5)
      expect(grp[1].getAltitude()).toEqual(20.5)
    })

    it("adds_routing_each_aircraft", () => {
      const grp = new AircraftGroup({
        nContacts: 2,
      })
      const dest = new Point(20, 20)
      const myMock = jest.spyOn(Aircraft.prototype, "addRoutingPoint")
      grp.addRoutingPoint(dest)
      expect(grp.getNextRoutingPoint()).toEqual(dest)
      expect(myMock).toHaveBeenCalledTimes(2)
    })

    it("on_task", () => {
      const grp = new AircraftGroup({
        nContacts: 2,
      })
      grp.setTasking(new Tasking())
      expect(grp.isOnTask()).toEqual(true)
    })

    it("group_ac_type", () => {
      const grp = new AircraftGroup({ type: ACType.RPA })
      expect(grp.getType()).toEqual(ACType.RPA)
    })
  })
})
