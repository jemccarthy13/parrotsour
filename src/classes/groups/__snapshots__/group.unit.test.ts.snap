// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AircraftGroup draw_group draws_all_groups 1`] = `<canvas />`;
